pilot/pkg/serviceregistry/federation/controller.go

```diff
@@ -0,0 +1,463 @@
+			x++
+			if x%255 == 0 {
+				x++
+			}
+			if x >= maxIPs {
+				logger.Errorf("out of IPs to allocate for service entries")
+				return services
+			}
+			thirdOctet := x / 255
+			fourthOctet := x % 255
+			svc.AutoAllocatedAddress = fmt.Sprintf("240.241.%d.%d", thirdOctet, fourthOctet)
+		}
+	}
+	return services
+}
+
+// Services lists services
+func (c *Controller) Services() ([]*model.Service, error) {
+	c.storeLock.RLock()
+	defer c.storeLock.RUnlock()
+	return autoAllocateIPs(c.serviceStore), nil
+}
+
+// GetService retrieves a service by hostname if exists
+func (c *Controller) GetService(hostname host.Name) (*model.Service, error) {
+	var errs error
+	var out *model.Service
+	return out, errs
+}
+
+// NetworkGateways merges the service-based cross-network gateways from each registry.
+func (c *Controller) NetworkGateways() map[string][]*model.Gateway {
+	c.storeLock.RLock()
+	defer c.storeLock.RUnlock()
+	gws := map[string][]*model.Gateway{}
+	gws[c.networkName] = c.gatewayStore
+	return gws
+}
+
+// InstancesByPort retrieves instances for a service on a given port that match
+// any of the supplied labels. All instances match an empty label list.
+func (c *Controller) InstancesByPort(svc *model.Service, port int, labels labels.Collection) []*model.ServiceInstance {
+	instances := []*model.ServiceInstance{}
+	c.storeLock.RLock()
+	for _, instanceList := range c.instanceStore {
+		for _, instance := range instanceList {
+			if instance.Service == svc && instance.ServicePort.Port == port {
+				instances = append(instances, instance.DeepCopy())
+			}
+		}
+	}
+	c.storeLock.RUnlock()
+	return instances
+}
+
+// GetProxyServiceInstances lists service instances co-located with a given proxy
+func (c *Controller) GetProxyServiceInstances(node *model.Proxy) []*model.ServiceInstance {
+	return make([]*model.ServiceInstance, 0)
+}
+
+func (c *Controller) GetProxyWorkloadLabels(proxy *model.Proxy) labels.Collection {
+	return labels.Collection{}
+}
+
+// Run starts all the controllers
+func (c *Controller) Run(stop <-chan struct{}) {
+	eventCh := make(chan *federation.WatchEvent)
+	refreshTicker := time.NewTicker(c.resyncPeriod)
+	defer refreshTicker.Stop()
+	c.resync()
+	go func() {
+		for {
+			logger.Info("starting watch")
+			err := c.watch(eventCh, stop)
+			if err != nil {
+				logger.Errorf("watch failed: %s", err)
+				time.Sleep(c.backoffPolicy.NextBackOff())
+			} else {
+				return
+			}
+		}
+	}()
+	for {
+		select {
+		case <-stop:
+			logger.Info("Federation Controller terminated")
+			return
+		case e := <-eventCh:
+			logger.Debugf("watch event received: %s service %s", e.Action, e.Service.Name)
+			c.handleEvent(e)
+		case <-refreshTicker.C:
+			_ = c.resync()
+		}
+	}
+}
+
+func (c *Controller) handleEvent(e *federation.WatchEvent) {
+	if e.Service == nil {
+		checksum := c.resync()
+		if checksum != e.Checksum {
+			// this shouldn't happen
+			logger.Error("checksum mismatch after resync")
+		}
```

Comment:
I think there needs to be a `return` here.

Reply:
I added a return below this line

Reply:
yep